### DISCUSSION TITLE
Code listing with line numbers and caption

### DISCUSSION THREAD

**Discussion by jpmorr** [quarto-dev/quarto-cli#2189] [Q&A] [answered]
When I'm adding example code to a book project, how do I specify the code block to have a caption and to be formatted with line numbers while still having the relevant syntax highlighting? For example how would I create the following listing? Treating it like a table doesn't seem to work. Maybe it's not currently possible to do something like this with quarto?

```
```{python}
def sum(a, b):
    sum = a + b
    print(f'Sum: {sum}`)
    return sum
``` : Pointless implementation of a sum function {#lst-sumfunc}

```

**Comment by cscheid** [comment, marked as answer]
````
---
code-line-numbers: true
---

::: {#fig-sumfunc}

```python
def sum(a, b):
    sum = a + b
    print(f'Sum: {sum}')
    return sum
sum(1,2)
```

Pointless implementation of a sum function

:::
````

This is documented under https://quarto.org/docs/authoring/figures.html and https://quarto.org/docs/output-formats/html-code.html#highlighting

**Reply by jpmorr** [reply]
@cscheid That pretty much does what I want. However, that includes the caption/reference as **Figure**. Obviously if I change `#fig` to something else such as `eq` or `tbl` then the caption will change according. Is there any label that refers to code such as a `listing`. I've tried `lst` and that doesn't work. Or maybe even a custom caption label?

**Reply by jpmorr** [reply]
@cscheid Actually I found exactly what I wanted in the documentation: [Code Listings](https://quarto.org/docs/authoring/cross-references.html#code-listings ).  Only downside is you seem to lose the ability to specify the `filename` in a neat manner when using the list caption option `lst-cap="Customers Query"`.

```
{#lst-prefs .python lst-cap="Sum Function" filename=sum.py}
  def sum(a, b):
      sum = a + b
      print(f'Sum: {sum}`)
      return sum
```  

![image](https://user-images.githubusercontent.com/10863809/187267210-4485f8ba-9913-4553-a16c-75a5f60ec021.png)

I think the placement of the listing below the filename may be a bug?

**Reply by cscheid** [reply]
Yeah, that caption location looks like a bug to me too. It should either be above `sum.py` (when `fig-cap-location: top`) or below the listing (when `fig-cap-location: bottom`).

**Reply by cscheid** [reply]
In addition, it neither respects `fig-cap-location`, nor do we have a `lst-cap-location` option. Another bug...

**Reply by drriguz** [reply]
Still the caption is placed below the filename, any workaround?

**Reply by cscheid** [reply]
@drriguz This should be [fixed on 1.3](https://github.com/quarto-dev/quarto-cli/issues/2195) prereleases, which you can [download here](https://quarto.org/docs/download/prerelease).

**Reply by drriguz** [reply]
Thanks, I'll try it!

**Comment by AmeliaMN** [comment]
I posted on the [Posit community](https://community.rstudio.com/t/lst-cap-and-lst-label-in-quarto/157173) before being reminded this might be the right place to ask. I can't figure out how to use `lst-cap` and `lst-label`. Here's a minimal document that isn't behaving how I expected:

````
---
title: "Untitled"
format: html
editor: visual
---

```{r}
#| lst-label: lst-code
#| lst-cap: "Some R code"
a <- 1
a+1
```

I'd like to be able to reference the code chunk @lst-code.
````

**Reply by baptiste** [reply]
Seems to be tracked at https://github.com/quarto-dev/quarto-cli/issues/1580

**Reply by AmeliaMN** [reply]
Thanks!

**Reply by AmeliaMN** [reply]
Ah I figured it out-- the inline options are actually just a display code block. Doh!

**Comment by AmeliaMN** [comment]
I've been able to make it work by using inline chunk options, like so:

````
---
title: "Untitled"
format: pdf
editor: visual
---

``` {#lst-code .r lst-cap="Some R code"}
a <- 1
a+1
```

I'd like to be able to reference the code chunk @lst-code.
````

So perhaps the issue is with the in-chunk options?
